### PR TITLE
rpc: add path to gethdkey

### PR DESCRIPTION
**Pull request by Sjors**
Depends on #26728.

This PR expands the wallet RPC `gethdkey` introduced in #26728. It takes a BIP32 path as argument and returns the xpub, along with the master key fingerprint.

To test this PR:

* create a regular descriptor wallet
* call `getxpub m/86h/0h/0h`
* call `listdescriptors`, compare the xpub in the `tr()` descriptor with the previous step

## Bigger picture

This paves the way for using Bitcoin Core as one signer in a multisig setup. It simplifies the proposed flow in #22067.

The eventual flow would be:
1. Create a blank wallet with a seed (either with no descriptors, or where its single key descriptors are not `active`, so they don't get used when calling `getnewaddress`). 
2. Call `getxpub m/87h/0h/0h` (as suggested in [BIP 87](https://github.com/bitcoin/bips/blob/master/bip-0087.mediawiki))
3. (Manually, with Specter or with a simple Python utility - TBD): craft a multisig descriptor containing this xpub
4. Call `importdescriptors` which will allow the import if its own fingerprint is recognized (and after checking the xpub)
5. The usual flow with `getnewaddress`, `send` and `walletprocesspsbt` (and their GUI equivalents)

This PR makes step (2) possible.

Step (1) would require a followup, because blank wallets currently can't have a seed. Step 4 would also require a followup,  so that `importdescriptors` treats an `xpub` as if it was an `xpriv`, after checking that it can derive the xpub from the seed (that matches the fingerprint)

On the GUI side a wizard could perform the above steps and guide the user, where they either:
* export their own xpub and then import a descriptor from somewhere else (performing some sanity checks); or
* import xpub(s) from other wallets, compose a multisig and get the descriptors for export; or
* get xpub(s) from connected hardware devices and configure a multisig (with or without a key on their machine)

### CODE REVIEW

**Comment by jonatack** [comment]
Thanks for working on this!

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3258,3 +3258,34 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
...
+        return false;
+    }
+
+    // TODO: avoid needing a pre-existing ScriptPubKeyMan for obtaining the HD seed
+    auto spk_man = GetScriptPubKeyMan(OutputType::BECH32, true);
+    if (!spk_man) return false;
+    auto desc_spk_man = dynamic_cast<DescriptorScriptPubKeyMan*>(spk_man);
+    CKey seed = desc_spk_man->GetKeys().begin()->second;
+    CExtKey master_key;
+    master_key.SetSeed(seed.begin(), seed.size());
```
Comment (Sjors) [inline src/wallet/wallet.cpp:3277]: In addition to being an ugly hack, this is also incorrect and does not actually produce the correct master key.

**Comment by Rspigler** [comment]
Thanks so much for continuing to work on this! Love the use of BIP 87 :)

> Call `importdescriptors` which will allow the import if its own fingerprint is recognized (and after checking the xpub)

Yes, important to check the xpub also, not just the fingerprint, which is easy to spoof

> - export their own xpub and then import a descriptor...
> - import xpub(s) from other wallets, compose a multisig...
> - get xpub(s) from connected hardware devices...

I agree that the above 3 encompasses all use cases/needs for Bitcoin Core being a coordinator and a signer in a multisig setup.

> `importdescriptors` should treat an xpub as if it was an xpriv, after checking that it can derive the xpub from the seed (that matches the fingerprint); it needs to copy the seed from a sibling descriptor??

Can you explain this to me more?

As for the blank wallets, why do we need to be working from blank wallets, and if xprivs are derived from seeds, how can a blank wallet have a seed?

**Comment by Sjors** [comment]
It depends on how we want to define `blank`. It would be empty except for a seed, so without any descriptors in it.

If a wallet has a seed and you call `importdescriptors` on it with some multisig descriptor, it should check the fingerprint(s) and path(s) mentioned in that descriptor. It should then derive the xpriv & xpub for that path, from the wallet seed, and check if it matches the descriptor `xpub`. In that case, she wallet should replace that xpub with the xpriv, as if the user provided it upon import. That way, if you want to sign a transaction, the wallet knows it can provide one of the signatures.

**Comment by Rspigler** [comment]
Thank you, makes a lot of sense!

And what about sibling descriptors?

File: test/functional/wallet_getxpub.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def set_test_params(self):
+        self.num_nodes = 2
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+        self.skip_if_no_sqlite()
+
+    def run_test(self):
+        self.nodes[0].createwallet(wallet_name="w1")
+        w1 = self.nodes[0].get_wallet_rpc("w1")
```
Comment (lsilva01) [inline test/functional/wallet_getxpub.py:25]: With a wallet different from the default one, I receive the error when sending to w2 address.
`test_framework.authproxy.JSONRPCException: Insufficient funds (-6)`

With the change below, it worked.

```suggestion
        w1 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
```

File: test/functional/wallet_getxpub.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                                    ])
+        assert(res[0]['success'] and res[1]['success'])
+
+        self.log.info("Fund xpub")
+        # TODO: if createwallet supports setting a seed without descriptors,
+        #       use that here instead. This allows replacing deriveaddresses
+        #       with getnewaddress.
+        w2_address = self.nodes[1].deriveaddresses(descsum_create(desc_receive), 0)[0]
+        w1.sendtoaddress(w2_address, 1)
+        self.nodes[0].generate(1)
```
Comment (lsilva01) [inline test/functional/wallet_getxpub.py:57]: I think `self.sync_all()` is necessary for w2 to know that it received money.

```suggestion
        self.nodes[0].generate(1)
        self.sync_all()
```

File: test/functional/wallet_getxpub.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        self.log.info("Spend from xpub")
+        w1_address = self.nodes[0].getnewaddress()
+        w2.sendtoaddress(w1_address, 0.1)
```
[the remark this answers is not in the document]
Comment (lsilva01) [reply test/functional/wallet_getxpub.py:61]: Just confirming that I got the error: `test_framework.authproxy.JSONRPCException: Signing transaction failed (-6)`

**Review by lsilva01 — COMMENTED** [review]
Code Review ACK https://github.com/bitcoin/bitcoin/pull/22341/commits/1ff7a25d5af28481537f146632f57a3f20f53420

**Comment by mjdietzx** [comment]
Concept ACK

I'll be sure to update the tests/docs in #22067 to use this when this PR gets merged

**Comment by fjahr** [comment]
Concept ACK

**Comment by Zero-1729** [comment]
Concept ACK

**Comment by Sjors** [comment]
Ok, this is now ready for review.

> I'll be sure to update the tests/docs in #22067 to use this when this PR gets merged

I might also take a stab at this, once this PR (and its ancestor) is rebased to include it.

**Comment by Sjors** [comment]
Rebased and updated the multisig descriptor test and example documentation to use `getxpub`. Unfortunately we still have to reuse existing wallet keys, e.g. `m/44h/1h/0h`, because `importdescriptors` can't derive private keys yet.

**Comment by JeremyRubin** [comment]
concept ACK this + using core as a signer like this.


One also cool idea would be is if, during PSBT signing code, we temporarily treat all descriptors there as "ours"?

**Comment by Sjors** [comment]
@JeremyRubin afaik descriptors are always treated as "ours", both in private key wallets and watch-only wallets. The distinction between watch-only and regular addresses was dropped with the introduction descriptor wallets.

**Comment by mjdietzx** [comment]
@Sjors in the description of this PR you say "The current draft implementation is not usable." what specifically are you referring to by that?

**Comment by Sjors** [comment]
@mjdietzx I think that refers to an older version of this PR that used a hideous hack, but the current code should be fine to review. I deleted that remark for the description.

**Comment by mjdietzx** [comment]
Code review ACK 8042053d0ddd662f654207d6554fb8f1e806f729

**Review by mjdietzx — COMMENTED** [review]
Tested ACK 8042053d0ddd662f654207d6554fb8f1e806f729

I went through a variety of edge-cases checking that `getxpub` works as expected for error cases in various test wallets I have setup

File: doc/descriptors.md [hunk 5]
```diff
@@ -155,8 +155,8 @@ The basic steps are:
 
   1. Every participant generates an xpub. The most straightforward way is to create a new descriptor wallet which we will refer to as
      the participant's signer wallet. Avoid reusing this wallet for any purpose other than signing transactions from the
-     corresponding multisig we are about to create. Hint: extract the wallet's xpubs using `listdescriptors` and pick the one from the
-     `pkh` descriptor since it's least likely to be accidentally reused (legacy addresses)
+     corresponding multisig we are about to create. Hint: extract the wallet's xpubs using `getxpub m/44h/0h/0h` (`m/44h/1h/0h` for testnet and signet). This reuses keys from the
```
[the remark this answers is not in the document]
Comment (Sjors) [reply doc/descriptors.md:158]: If using `h` doesn't work that's a bug. I'll look into that (probably after the next rebase).

**Comment by Rspigler** [comment]
Tested commit 0ecd8a338e7f015e711cdc05ba005cdd322545e1

Got this error when running `make`:

>   CXX      wallet/libbitcoin_wallet_a-keyman.o
> In file included from ./script/signingprovider.h:11,
>                  from ./wallet/crypter.h:10,
>                  from wallet/keyman.cpp:9:
> ./script/keyorigin.h: In member function ‘std::optional<std::pair<CExtPubKey, KeyOriginInfo> > wallet::KeyManager::GetExtPubKey(std::vector<unsigned int>) const’:
> ./script/keyorigin.h:11:8: warning: ‘origin’ may be used uninitialized in this function [-Wmaybe-uninitialized]
>    11 | struct KeyOriginInfo
>       |        ^~~~~~~~~~~~~
> wallet/keyman.cpp:90:19: note: ‘origin’ was declared here
>    90 |     KeyOriginInfo origin;
>       |                   ^~~~~~
>   CXX      wallet/libbitcoin_wallet_a-load.o
> 

Still was able to compile

Tests all pass

Create a wallet with private keys disabled

1. Create a blank wallet with a seed (single key descriptor that is not active) as described in OP.

`importdescriptors "[{ \"desc\": \"pk(0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798)#gn28ywm7\", \"active\": false, \"timestamp\": \"now\" }]"`

> [
>   {
>     "success": true
>   }
> ]

2. Call `getxpub m/87h/0h/0h`

`getxpub m/87h/0h/0h`

>This wallet does not have an active master extended key (code -4)

Hm, ok - I'll repeat with a descriptor wallet, but won't be able to test the `getnewaddress` flow

1. Create a regular descriptor wallet
2. Call `getxpub m/87h/0h/0h`

`getxpub m/87h/0h/0h`

> {
>   "xpub": "xpub6BxZHpQ2UJZML59GZRPirFsWFJDqcYn4jcM6zm5qHUZnvuCC7xmb86meQywCbKuGEiuh3XvrvG1bhRiQDsYGmGRhFkMRNqYnBVtJYPCVt43",
>   "fingerprint": "5d340155",
>   "origin": "[5d340155/87'/0'/0']"
> }

Test importdescriptors:

`importdescriptors "[{ \"desc\": \"wpkh([5d340155/87'/0'/0]xpub6BxZHpQ2UJZML59GZRPirFsWFJDqcYn4jcM6zm5qHUZnvuCC7xmb86meQywCbKuGEiuh3XvrvG1bhRiQDsYGmGRhFkMRNqYnBVtJYPCVt43/0/*)#zc6l5c3g\", \"timestamp\": \"now\" }]"`

> [
>   {
>     "success": false,
>     "error": {
>       "code": -4,
>       "message": "Cannot import descriptor without private keys to a wallet with private keys enabled"
>     },
>     "warnings": [
>       "Range not given, using default keypool range"
>     ]
>   }
> ]

Not sure how to proceed from here

**Comment by Sjors** [comment]
> Got this error when running make:

@Rspigler are you sure that warning is not related to #23417?

I clarified the description a bit: the steps there don't currently work.

**Comment by Rspigler** [comment]
I am able to compile and run #23417 without issue.

But still get this error when compiling this PR:

> CXX      wallet/libbitcoin_wallet_a-keyman.o
> In file included from ./script/signingprovider.h:11,
>                  from ./wallet/crypter.h:10,
>                  from wallet/keyman.cpp:9:
> ./script/keyorigin.h: In member function ‘std::optional<std::pair<CExtPubKey, KeyOriginInfo> > wallet::KeyManager::GetExtPubKey(std::vector<unsigned int>) const’:
> ./script/keyorigin.h:11:8: warning: ‘origin’ may be used uninitialized in this function [-Wmaybe-uninitialized]
>    11 | struct KeyOriginInfo
>       |        ^~~~~~~~~~~~~
> wallet/keyman.cpp:90:19: note: ‘origin’ was declared here
>    90 |     KeyOriginInfo origin;
>       |                   ^~~~~~
>   CXX      wallet/libbitcoin_wallet_a-load.o
> 

I understand the PR better now, thank you for clarifying the description.

create a regular descriptor wallet
`listdescriptors`:
...(listing taproot descriptors only):

> 
> {
>       "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/0/*)#p5us77gk",
>       "timestamp": 1657730232,
>       "active": true,
>       "internal": false,
>       "range": [
>         0,
>         999
>       ],
>       "next": 0
>     },
>     
>     {
>       "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/1/*)#sqe3rtcw",
>       "timestamp": 1657730234,
>       "active": true,
>       "internal": true,
>       "range": [
>         0,
>         999
>       ],
>       "next": 0
>     }


`getxpub "m/86'/0'/0'"`

> {
>   "xpub": "xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk",
>   "fingerprint": "3e89a5ab",
>   "origin": "[3e89a5ab/86'/0'/0']"
> }


'listdescriptors'
(listing taproot only)

{
      "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/0/*)#p5us77gk",
      "timestamp": 1657730232,
      "active": true,
      "internal": false,
      "range": [
        0,
        999
      ],
      "next": 0
    },
    
    
    

  {
      "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/1/*)#sqe3rtcw",
      "timestamp": 1657730234,
      "active": true,
      "internal": true,
      "range": [
        0,
        999
      ],
      "next": 0
    }


The taproot descriptors are the same 

:heavy_check_mark: 

tACK commit 0ecd8a338e7f015e711cdc05ba005cdd322545e1

**Comment by Sjors** [comment]
> But still get this error when compiling this PR:

Ok, I'll look into that, probably after the next rebase.

File: doc/descriptors.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply doc/descriptors.md:162]: The test cover both the use of `'` and `h`. You're not allowed to mix them in the same descriptor though.

**Comment by Sjors** [comment]
Rebased. @Rspigler the uninitialized variable warning should be gone now.

**Comment by Rspigler** [comment]
tACK commit f6133f093e3e305b305506b68499e1ac6cde726b

Yes, it is all good now

**Comment by Sjors** [comment]
(trivial rebase, but not worth reviewing yet because there will be another one soon)

**Comment by Sjors** [comment]
Rebased and fixed `nodiscard` warning introduced in #25642.

**Review by w0xlt — COMMENTED** [review]
Approach ACK

**Review by JurbinM — APPROVED** [review]

File: src/rpc/util.cpp [hunk 6]
```diff
@@ -1092,6 +1093,15 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
     return ret;
 }
 
+std::vector<uint32_t> ParsePathBIP32(const std::string path)
```
Comment (aureleoules) [inline src/rpc/util.cpp:1096]:
```suggestion
std::vector<uint32_t> ParsePathBIP32(const std::string& path)
```

File: src/wallet/keyman.cpp [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+std::optional<CExtPubKey> KeyManager::GetActiveHDPubKey() const
+{
+    if (!m_active_xpub.pubkey.IsValid()) {
+        return std::nullopt;
+    }
+    return m_active_xpub;
+}
+
+std::optional<std::pair<CExtPubKey, KeyOriginInfo>> KeyManager::GetExtPubKey(const std::vector<uint32_t> path) const {
```
Comment (aureleoules) [inline src/wallet/keyman.cpp:87]:
```suggestion
std::optional<std::pair<CExtPubKey, KeyOriginInfo>> KeyManager::GetExtPubKey(const std::vector<uint32_t>& path) const {
```

**Review by aureleoules — COMMENTED** [review]
Concept ACK & LGTM

**Comment by Sjors** [comment]
Rebased on top of #26728 instead of #23417.

**Comment by jonatack** [comment]
Re-concept ACK (not sure if it's ready for review).

**Comment by Sjors** [comment]
@jonatack it was a somewhat sloppy rebase. I also need to review the new base PR. Will mark as draft.

**Comment by Sjors** [comment]
Can't reproduce the CI failure, but will try again once the upstream PR is merged or rebased past #26076.

**Comment by S3RK** [comment]
The description now says the PR depends on itself :-)

**Comment by Sjors** [comment]
I plan to re-work this on top of  #29130 with a similar interface to `createwalletdescriptor`. That is, if you have a regular wallet with normal descriptors, it will Just Work(tm). Otherwise you need to specify which master key to use.

Initially I'll make a separate PR. If #26728 is closed in favor of the new approach in #29130 then this PR can be closed too.

**Comment by ryanofsky** [comment]
> Initially I'll make a separate PR. If #26728 is closed in favor of the new approach in #29130 then this PR can be closed too.

It seems like you could just rebase this on top of #29130, since the description of the PR and commits aren't going to change that much, they will just be based on #29130 instead of #26728. Either way seems fine, though.

**Comment by Sjors** [comment]
@ryanofsky building on #29130 introduces a new argument to specify which master key you intend to use.

**Comment by Sjors** [comment]
Given that #26728 was closed, I'll just modify this PR to build on top of #29130.

(bit busy with stratum v2 stuff, but it will happen)

**Comment by Sjors** [comment]
This seems more involved than a simple rebase. #29130 changes `gethdkey` to `gethdkeys` making it a less obvious choice to add a `path` argument to. We probably need a new RPC that's more similar to `createwalletdescriptor` in how it's called.

I probably won't have time for this in the near future, so I'm closing this as up for grabs.

**Pull request closed without merging**